proposals/p3762.md

```diff
@@ -0,0 +1,828 @@
+the `impl` is not defined locally. Note that an entity declared in the `api` may
+still not get a compile-time diagnostic unless the compiler is told it's seeing
+_all_ available `impl` files.
+
+#### Type scopes may contain both a forward declaration and definition
+
+The combination of a forward declaration and a definition is allowed in type
+scopes.
+
+For example:
+
+```carbon
+class C {
+  class D;
+
+  fn F() -> D;
+
+  class D {
+    fn G() -> Self { return C.F(); }
+
+    var x: i32;
+  }
+
+  fn F() -> D { return {.x = 42}; }
+}
+```
+
+This is necessary because type bodies are not automatically moved out-of-line,
+unlike function bodies.
+
+### Merging `extern` declarations
+
+When merging imported declarations from different libraries, multiple
+declarations for the same entity may be found. Only one declaration is retained
+for semantic checking and diagnostics. If available, a non-`extern` declaration
+(potentially with a definition) is used; otherwise, the `extern` declaration
+from the earliest import is used.
+
+For a new declaration in the importing file, an imported `extern` declaration is
+always superseded. It is invalid to use the imported declaration before the new
+declaration. It is considered the same entity for type checking, but diagnostics
+should use the local definition.
+
+For a new declaration in the importing file, an imported non-`extern`
+declaration will:
+
+-   If the new declaration is `extern`, the imported non-`extern` is superseded
+    as above. It is invalid to use the imported declaration before the new
+    declaration.
+-   If the new declaration is non-`extern`, the imported declaration must be
+    from the same library.
+    -   The `api` may forward declare when the `impl` defines, or the `impl` may
+        repeat a forward declaration and also define.
+
+### Using `extern` declarations in `extern impl`
+
+An `extern` type may only be used in the type structure of an `impl` if it's an
+`extern impl`. They are incompatible with non-`extern` `impl`s.
+
+Consider two libraries, one defining `A` and declaring `B` as `extern`, and the
+other defining `B` and declaring `A` as `extern`. Neither should be able to
+define an `impl` involving both `A` and `B`, otherwise both could.
+
+### Modifier keywords
+
+When considering various modifiers on a forward declaration versus definition:
+
+-   `extern` is only valid on a forward declaration. Rules are detailed above.
+-   `extend` in `extend impl` is only on the declaration in the class body
+    (whether that is a forward declaration or definition), as described at
+    [Forward `impl` declaration](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#forward-impl-declaration).
+-   Other class, impl, and interface modifiers (`abstract`, `base`, `final`)
+    exist only on the definition, not on the forward declaration.
+-   Function modifiers (`impl`, `virtual`, `default`, `abstract`, `base`,
```

Comment:
I think the `base` modifier is only permitted on class declarations (and as an introducer for `base: B`).

```suggestion
-   Function modifiers (`impl`, `virtual`, `default`, `abstract`,
```